roles/proxy_mysql/tasks/main.yml

```diff
@@ -4,12 +4,34 @@
     src: "create_db.sql"
     dest: /tmp
 
+- name: Copy foreign keys file
+  copy:
+    src: "fkeys.sql"
+    dest: /tmp
+
+- name: Copy views file
+  copy:
+    src: "views.sql"
+    dest: /tmp
+
 - name: create database for CoManage-Proxy linkup
   mysql_db:
     name: "{{ comanage_proxy_db_name }}"
     state: import
     target: /tmp/create_db.sql
 
+- name: create foreign key relations for CoManage-Proxy linkup
+  mysql_db:
+    name: "{{ comanage_proxy_db_name }}"
+    state: import
+    target: /tmp/fkeys.sql
+
+- name: create views for CoManage-Proxy linkup
+  mysql_db:
+    name: "{{ comanage_proxy_db_name }}"
+    state: import
+    target: /tmp/views.sql
+
```

Comment:
These two tasks don't seem to be idempotent (i.e., they return a s status "changed" in the second invocation).  Could you please fix that?

Reply:
That change is not trivial. The idempotency is inside the scripts using the 'if not exists' construct. The alternative would be something like checking for the existance of views and keys in a separate construct, register that status and then execute the script based on that status. That would mean two separate tasks at least for each key and each view.
Furthermore, the 'create_db.sql' script is run on each deploy as well and will, in the same vein, also report 'changed'. So that would need adjustment as well.

The mysql_db module documentation states 'This module is not idempotent when state is import, and will import the dump file each time if run more than once.' There does not seem to be an easy way around this, which will force implementation of a complicated, hard-to-maintain task list with other potential deployment issues and other 'undocumented features'.

Reply:
The problem is that the non-idempotent tasks complicate the deployment, so I would like to get rid of them for as much as possible.  In this case, we could use a sql statement like this:
```sql
DELIMITER $
BEGIN NOT ATOMIC
  IF ( select count(*) from information_schema.views where TABLE_SCHEMA='cotosa' and TABLE_NAME='view_people_services')=1 
  THEN
    SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT='view exists';
  ELSE
    create view view_people_services as SELECT              
      ps.zone_service_id, s.metadata, s.`attributes` as service_attributes,
      ps.zone_person_id, p.uid, p.`attributes` as person_attributes,p.co_id
      FROM cotosa.zone_person_zone_service as ps
      LEFT JOIN zone_services as s on s.id=ps.zone_service_id
      LEFT JOIN zone_people   as p on p.id=ps.zone_person_id
      ;
  END IF;
END $
DELIMITER ;
```
Then we can import this using `command: mariadb` and check stdout for `view exists`.

For the constraint, we can take a similar approach, but use a query like
```sql
select count(*) from information_schema.key_column_usage where constraint_name='fk_zone_person_id';
```

That's still not optimal, but shouldn't be hard to maintain and fixes the deployment issues.

Alternatively, we could implement a similar functionality in a custom ansible module (along the line s of this: https://github.com/zauberpony/ansible-mysql-query/blob/master/library/mysql_query.py), but that doesn't really seem worth the effort.